File: trunk/canvas4z/src/main/resources/web/js/canvas/Canvas.js
/* Canvas.js

	Purpose:
		
	Description:
		
	History:
		May 12, 2010 3:17:24 PM , Created by simon

Copyright (C) 2010 Potix Corporation. All Rights Reserved.

*/
(function () {
	
	function _getRectFromBound(wgt, bnd) {
		var cvs = wgt._cvs;
		if (!bnd)
			return [0, 0, cvs.width, cvs.height];
		var x0 = bnd.x0 || 0,
			y0 = bnd.y0 || 0,
			w = (bnd.x1 || cvs.width) - x0,
			h = (bnd.y1 || cvs.height) - y0;
		return [x0, y0, w, h];
	}
	function _getTargetIndex(wgt, evt) {
		var n = wgt.$n(),
			x = evt.data.pageX - n.offsetLeft,
			y = evt.data.pageY - n.offsetTop;
		for (var drws = wgt._drwbls, i = drws.length, d; i--;) 
			if ((d = drws[i]) && d.contains(x, y))
				return i;
		return -1;
	}
	
/**
 * The ZK component corresponding to HTML 5 Canvas.
 * While HTML 5 Canvas is a command-based DOM object that allows user to draw
 * items on a surface, ZK Canvas maintains a list of drawable items and allow
 * user to operate the list by adding, removing, updating, replacing the 
 * elements. The changes will be reflected on the client side upon these
 * operations.
 * 
 * <p>Default {@link #getZclass}: z-canvas.
 */
canvas.Canvas = zk.$extends(zul.Widget, {
	_cvs: null,
	_ctx: null,
	_drwbls: [],
	_states: [],
	_hvi: -1,
	_sldi: -1,
	
	// extended drawing states
	_drwTp: "fill",
	_drwTpBak: [],
	_txtMxW: -1,
	_txtMxWBak: [],
	
	// TODO: rerender upon resize
	
	setDrwngs: function (drwsJSON) {
		this._drwbls = canvas.Drawable.createAll(drwsJSON);
		if (this.desktop)
			this._repaint();
	},
	setAddAll: function (drwsJSON) { // called by server
		this.addAll(canvas.Drawable.createAll(drwsJSON));
	},
	/**
	 * Adds a Drawable to canvas.
	 */
	add: function (drw) {
		this._paint(drw);
		this._drwbls.push(drw);
	},
	addAll: function (drws) {
		for (var i = 0, len = drws.length; i < len; i++)
			this.add(drws[i]);
	},
	setRemove: function (index) { // called by server
		this.remove(index);
	},
	/**
	 * Removes the Drawable at specific index.
	 */
	remove: function (index) { // TODO: remove batch
		var drw = this._drwbls.splice(index, 1);
		this._repaint(drw.getBound_());
		return drw;
	},
	setInsert: function (idrwsJSON) { // called by server
		var idrws = canvas.Drawable.createAll(idrwsJSON);
		this.insert(idrws.i, idrws.drws);
	},
	/**
	 * Inserts a Drawable at specific index.
	 */
	insert: function (index, drws) {
		var drwbls = this._drwbls; 
		drwbls.splice.apply(drwbls, [index, 0].concat(drws));
		this._repaint(/*drw.getBound_()*/); // TODO: get all bounds
	},
	setReplace: function (idrwJSON) { // called by server
		var idrw = canvas.Drawable.create(idrwJSON);
		this.replace(idrw.i, idrw.drw);
	},
	/**
	 * Replace the Drawable at specific index.
	 */
	replace: function (index, drw) { // TODO: replace batch
		var removed = this._drwbls.splice(index, 1, drw),
			bnd0 = removed[0].getBound_(),
			bnd1 = drw.getBound_();
		this._repaint(canvas.Drawable._joinBounds(bnd0, bnd1));
		return removed[0];
	},
	setClear: function () { // called by server
		this.clear();
	},
	/**
	 * Remove all Drawables.
	 */
	clear: function () {
		this._drwbls = [];
		this._clearCanvas();
	},
	/*
	// experimental, does not work on Chrome and IE
	saveAsPNG: function(width, height){
		if(!width) width = this._cvs.width;
		if(!height) height = this._cvs.height;
		Canvas2Image.saveAsPNG(this._cvs, false, width, height);
	},
	*/
	
	
	
	// private //
	_clearCanvas: function (bnd) {
		var r = _getRectFromBound(this, bnd);
		this._ctx.clearRect(r[0], r[1], r[2], r[3]);
	},
	_repaint: function (bnd) { // TODO: setTimeout
		this._clearCanvas(bnd);
		this._applyBound(bnd);
		for (var i = 0, drws = this._drwbls, len = drws.length; i < len; i++)
			this._paint(drws[i]);
		this._unapplyBound();
	},
	_paint: function (drw) {
		drw.paint_(this); // TODO: rethink design
		/*
		drw.applyState_(this);
		drw.paintObj_(this, bnd);
		drw.unapplyState_(this);
		*/
	},
	_applyBound: function (bnd) {
		var ctx = this._ctx;
		ctx.save();
		if (bnd) {
			var cvs = this._cvs,
				r = _getRectFromBound(this, bnd);
			ctx.beginPath();
			ctx.rect(r[0], r[1], r[2], r[3]);
			ctx.clip();
		}
	},
	_unapplyBound: function () {
		this._ctx.restore();
	},
	// state management helper //
	_applyLocalState: function (st) {
		// save current global state on DOM canvas context
		this._txtMxWBak.push(this._txtMxW);
		this._drwTpBak.push(this._drwTp);
		this._ctx.save();
		// apply local state to context
		this._setDOMContextState(st);
	},
	_setDOMContextState: function (st) {
		var ctx = this._ctx;
		// drawing type is NOT a part of DOM Canvas state
		if(st.dwtp) 
			this._drwTp = st.dwtp;
		if(st.trns) { // transformation
			var trns = st.trns;
			ctx.setTransform(trns[0], trns[1], trns[2], trns[3], trns[4], trns[5]);
		}
		if(st.clp) { // clipping
			canvas.Path.doPath(this, st.clp.obj);
			ctx.clip();
			ctx.beginPath();
		}
		if(st.strk) ctx.strokeStyle   = st.strk;
		if(st.fil)  ctx.fillStyle     = st.fil;
		if(st.alfa) ctx.globalAlpha   = st.alfa;
		if(st.lnw)  ctx.lineWidth     = st.lnw;
		if(st.lncp) ctx.lineCap       = st.lncp;
		if(st.lnj)  ctx.lineJoin      = st.lnj;
		if(st.mtr)  ctx.miterLimit    = st.mtr;
		if(st.shx)  ctx.shadowOffsetX = st.shx;
		if(st.shy)  ctx.shadowOffsetY = st.shy;
		if(st.shb)  ctx.shadowBlur    = st.shb;
		if(st.shc)  ctx.shadowColor   = st.shc;
		if(st.cmp)  ctx.globalCompositeOperation = st.cmp;
		if(st.fnt)  ctx.font          = st.fnt;
		if(st.txal) ctx.textAlign     = st.txal;
		if(st.txbl) ctx.textBaseline  = st.txbl;
		// maxWidth is not a part of DOM Canvas state
		if(st.txmw) this._txtMxW = st.txmw;
	},
	_unapplyLocalState: function () {
		// restore global state
		this._txtMxW = this._txtMxWBak.pop();
		this._drwTp = this._drwTpBak.pop();
		this._ctx.restore();
	},
	//@Override
	bind_: function () {
		this.$supers("bind_", arguments);
		zWatch.listen({onSize: this, onShow: this});
		
		this._cvs = document.createElement("canvas");
		this._init();
		this._cvs.id = this.uuid + '-cnt';
		
		this._ctx = this._cvs.getContext("2d");
		jq(this.$n()).append(this._cvs);
		
		this._repaint();
	},
	unbind_: function () {
		zWatch.unlisten({onSize: this, onShow: this});
		this.$supers("unbind_", arguments);
	},
	_init: function () {
		var n = this.$n(),
			w = n.offsetWidth,
			h = n.offsetHeight,
			cvs = this._cvs;
		if (w) 
			cvs.width = zk.parseInt(w);
		if (h) 
			cvs.height = zk.parseInt(h);
		if (zk.ie < 9) 
			G_vmlCanvasManager.initElement(cvs);
	},
	onSize: _zkf = function () {
		this._init();
		this._repaint();
	},
	onShow: _zkf,
	setWidth: function () {
		this.$supers('setWidth', arguments);
		if (this.$n())
			this.onSize();
	},
	setHeight: function () {
		this.$supers('setHeight', arguments);
		if (this.$n())
			this.onSize(); // TODO: refine
	},
	doMouseMove_: function (evt) {
		var i = _getTargetIndex(this, evt),
			pi = this._hvi;
		if (pi != i) {
			this.fire('onTooltip', zk.copy(evt.data, {i: i, pi: pi}));
			this._hvi = i;
		}
		this.$supers('doMouseMove_', arguments);
	},
	doClick_: function (evt) {
		var i = _getTargetIndex(this, evt),
			pi = this._sldi; // TODO
		if (pi != i) {
			this.fire('onSelect', zk.copy(evt.data, {i: i, pi: pi}));
			this._sldi = i;
		}
		this.$supers('doClick_', arguments);
	},
	getZclass: function () {
		var zcs = this._zclass;
		return zcs != null ? zcs: "z-canvas";
	}
	
});

})();

File: 0.9.0-VisionSoftware/src/main/java/org/zkoss/canvas/Canvas.java
/* Canvas.java

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		May 12, 2010 3:16:36 PM , Created by simon
}}IS_NOTE

Copyright (C) 2010 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
}}IS_RIGHT
*/
package org.zkoss.canvas;

import java.awt.geom.Path2D;
import java.awt.geom.Rectangle2D;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.zkoss.canvas.drawable.CanvasSnapshot;
import org.zkoss.canvas.drawable.Drawable;
import org.zkoss.canvas.drawable.ImageSnapshot;
import org.zkoss.canvas.drawable.Path;
import org.zkoss.canvas.drawable.Rectangle;
import org.zkoss.canvas.drawable.Text;
import org.zkoss.json.JSONObject;
import org.zkoss.json.JSONValue;
import org.zkoss.zk.au.AuRequest;
import org.zkoss.zk.ui.event.Events;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.Image;
import org.zkoss.zul.impl.XulElement;

/**
 * The prototype component corresponding to HTML 5 Canvas.
 * While HTML 5 Canvas is a command-based DOM object that allows user to draw
 * items on a surface, This Canvas maintains a list of drawable items and allow
 * user to operate the list by adding, removing, updating, replacing the 
 * elements. The changes will be reflected on the client side upon these
 * operations.
 * 
 * @author simon
 *
 */
@SuppressWarnings("serial")
public class Canvas extends XulElement {
	
	private List<Drawable> _drawables;
	
	static {
		addClientEvent(Canvas.class, CanvasMouseEvent.ON_TOOLTIP, CE_REPEAT_IGNORE);
		addClientEvent(Canvas.class, CanvasMouseEvent.ON_SELECT, 0);
	}
	
	public Canvas() {
		_drawables = new ArrayList<Drawable>();
	}
	
	/**
	 * Return a list of all drawings in Canvas.
	 */
	public List<Drawable> getAllDrawables(){
		return Collections.unmodifiableList(_drawables);
	}
	
	/**
	 * Returns the drawing at position index. 
	 * @param index: drawings at 0 is the earliest drawing.
	 */
	public Drawable getDrawable(int index){
		return _drawables.get(index);
	}
	
	/**
	 * Return true if the list is empty.
	 */
	public boolean isEmpty(){
		return _drawables.isEmpty();
	}
	
	/**
	 * Returns the size of Drawable lists
	 */
	public int size(){
		return _drawables.size();
	}
	
	/**
	 * Adds the Drawable object to the end of the list.
	 */
	public void add(Drawable drawable){
		_drawables.add(cloneIfPossible(drawable));
		smartUpdate("add", drawable.toJSONString(), true);
	}
	
	/**
	 * Adds a Java 2D Path object.
	 */
	public void add(Path2D.Double path){
		add(new Path(path));
	}
	
	/**
	 * Adds a Java 2D Rectangle object.
	 */
	public void add(Rectangle2D.Double rectangle){
		add(new Rectangle(rectangle));
	}
	
	/**
	 * Adds a Text object.
	 */
	public void add(String text, double x, double y){
		add(new Text(text, x, y));
	}
	
	// TODO: delegate other image type
	/**
	 * Adds an Image snapshot
	 */
	public void add(Image image, double dx, double dy){
		add(new ImageSnapshot(image, dx, dy));
	}
	
	/**
	 * Adds an Image snapshot
	 */
	public void add(Image image, double dx, double dy, double dw, double dh){
		add(new ImageSnapshot(image, dx, dy, dw, dh));
	}
	
	/**
	 * Adds an Image snapshot
	 */
	public void add(Image image, double dx, double dy, double dw, double dh,
			double sx, double sy, double sw, double sh){
		add(new ImageSnapshot(image, dx, dy, dw, dh, sx, sy, sw, sh));
	}
	
	/**
	 * Adds a Canvas snapshot
	 */
	public void add(Canvas canvas, double dx, double dy){
		add(new CanvasSnapshot(canvas, dx, dy));
	}
	
	/**
	 * Adds a Canvas snapshot
	 */
	public void add(Canvas canvas, double dx, double dy, double dw, double dh){
		add(new CanvasSnapshot(canvas, dx, dy, dw, dh));
	}
	
	/**
	 * Adds a Canvas snapshot
	 */
	public void add(Canvas canvas, double dx, double dy, double dw, double dh,
			double sx, double sy, double sw, double sh){
		add(new CanvasSnapshot(canvas, dx, dy, dw, dh, sx, sy, sw, sh));
	}
	
	/**
	 * Removes the Drawable at specific index.
	 * @return The removed Drawable
	 */
	public Drawable remove(int index){
		Drawable removed = _drawables.remove(index);
		smartUpdate("remove", index, true);
		return removed;
	}
	
	/**
	 * Clears the Drawable list. The Canvas is also cleared as a result.
	 */
	public void clear(){
		_drawables.clear();
		smartUpdate("clear", null);
	}
	
	/**
	 * Inserts the Drawable at specific index
	 */
	@SuppressWarnings("unchecked")
	public void insert(int index, Drawable drawable){
		_drawables.add(index, cloneIfPossible(drawable));
		JSONObject args = new JSONObject();
		args.put("i", index);
		args.put("drw", drawable);
		smartUpdate("insert", args, true);
	}
	
	/**
	 * Inserts a Path.
	 */
	public void insert(int index, Path2D.Double path){
		insert(index, new Path(path));
	}
	
	/**
	 * Inserts a Rectangle.
	 */
	public void insert(int index, Rectangle2D.Double rectangle){
		insert(index, new Rectangle(rectangle));
	}
	
	/**
	 * Inserts a piece of Text.
	 */
	public void insert(int index, String text, double x, double y){
		insert(index, new Text(text, x, y));
	}
	
	/**
	 * Replace a Drawable at specific index.
	 * @return The replaced Drawable
	 */
	@SuppressWarnings("unchecked")
	public Drawable replace(int index, Drawable drawable){
		Drawable removed = _drawables.remove(index);
		_drawables.add(index, cloneIfPossible(drawable));
		JSONObject args = new JSONObject();
		args.put("i", index);
		args.put("drw", drawable);
		smartUpdate("replace", args, true);
		return removed;
	}
	
	/**
	 * Replaces a Path.
	 */
	public Drawable replace(int index, Path2D.Double path){
		return replace(index, new Path(path));
	}
	
	/**
	 * Replaces a Rectangle.
	 */
	public Drawable replace(int index, Rectangle2D.Double rectangle){
		return replace(index, new Rectangle(rectangle));
	}
	
	/**
	 * Replaces a piece of Text.
	 */
	public Drawable replace(int index, String text, double x, double y){
		return replace(index, new Text(text, x, y));
	}
	
	
	
	// helper //
	private static Drawable cloneIfPossible(Drawable drawable){
		/*
		try {
			return (Drawable) drawable.clone();
		} catch (CloneNotSupportedException e) {}
		*/
		return drawable;
	}
	
	// service //
	public void service(AuRequest request, boolean everError) {
		final String cmd = request.getCommand();
		if (cmd.equals(CanvasMouseEvent.ON_SELECT) || 
				cmd.equals(CanvasMouseEvent.ON_TOOLTIP)) {
			CanvasMouseEvent evt = CanvasMouseEvent.getEvent(request);
			Events.postEvent(evt);
		} else 
			super.service(request, everError);
	}
	
	// super //
	protected void renderProperties(ContentRenderer renderer) 
		throws IOException {
		
		super.renderProperties(renderer);
		render(renderer, "drwngs", JSONValue.toJSONString(_drawables));
		
	}
	
	public String getZclass() {
		return _zclass == null ? "z-canvas" : _zclass;
	}
	
}

